File: typescript/src/resolvers.ts
/**
 * SecretResolver interface and built-in implementations.
 *
 * See CONTRACT.md §Port/`SecretResolver protocol`. Each resolver handles one
 * or more schemes of `${scheme:value}`. Unknown schemes throw
 * {@link SecretSchemeUnsupported}; recognised-but-unresolvable values throw
 * {@link SecretRefUnresolved}. {@link ChainResolver} composes several.
 */

import { readFile } from "node:fs/promises";
import { isAbsolute, join } from "node:path";

import { SecretRefUnresolved, SecretSchemeUnsupported } from "./errors.js";

/** Resolve a `${scheme:value}` reference to a plain string. */
export interface SecretResolver {
  resolve(scheme: string, value: string): Promise<string>;
}

/** Resolve `${env:NAME}` from `process.env`. */
export class EnvResolver implements SecretResolver {
  static readonly scheme = "env";

  async resolve(scheme: string, value: string): Promise<string> {
    if (scheme !== EnvResolver.scheme) {
      throw new SecretSchemeUnsupported(
        `EnvResolver does not handle scheme '${scheme}'`,
      );
    }
    if (value.length === 0) {
      throw new SecretRefUnresolved("${env:} requires a variable name");
    }
    const resolved = process.env[value];
    if (resolved === undefined) {
      throw new SecretRefUnresolved(`env var not set: ${value}`);
    }
    return Promise.resolve(resolved);
  }
}

export interface FileResolverOptions {
  /** Directory used to resolve relative paths. Absolute paths ignore this. */
  baseDir?: string;
}

/**
 * Resolve `${file:path}` by reading the file's contents. Trailing whitespace
 * (including newlines) is stripped — matches the common "secret in a
 * single-line file" pattern used by Docker / Kubernetes.
 */
export class FileResolver implements SecretResolver {
  static readonly scheme = "file";
  private readonly baseDir: string | undefined;

  constructor(options: FileResolverOptions = {}) {
    this.baseDir = options.baseDir;
  }

  async resolve(scheme: string, value: string): Promise<string> {
    if (scheme !== FileResolver.scheme) {
      throw new SecretSchemeUnsupported(
        `FileResolver does not handle scheme '${scheme}'`,
      );
    }
    if (value.length === 0) {
      throw new SecretRefUnresolved("${file:} requires a path");
    }

    const path =
      !isAbsolute(value) && this.baseDir !== undefined
        ? join(this.baseDir, value)
        : value;

    try {
      const text = await readFile(path, "utf-8");
      return text.replace(/\s+$/, "");
    } catch (e) {
      const code = (e as { code?: string }).code;
      if (code === "ENOENT" || code === "ENOTDIR") {
        throw new SecretRefUnresolved(`secret file not found: ${path}`);
      }
      const msg = e instanceof Error ? e.message : String(e);
      throw new SecretRefUnresolved(`cannot read secret file ${path}: ${msg}`);
    }
  }
}

/**
 * Try each resolver in order; first one that handles the scheme wins. If
 * every resolver throws {@link SecretSchemeUnsupported}, the chain re-throws
 * {@link SecretSchemeUnsupported} naming the scheme. Any other exception
 * (notably {@link SecretRefUnresolved}) propagates immediately — scheme-
 * match wins the first handler.
 */
export class ChainResolver implements SecretResolver {
  private readonly resolvers: readonly SecretResolver[];

  constructor(...resolvers: SecretResolver[]) {
    this.resolvers = resolvers;
  }

  async resolve(scheme: string, value: string): Promise<string> {
    for (const r of this.resolvers) {
      try {
        return await r.resolve(scheme, value);
      } catch (e) {
        if (e instanceof SecretSchemeUnsupported) {
          continue;
        }
        throw e;
      }
    }
    throw new SecretSchemeUnsupported(
      `no resolver registered for scheme '${scheme}'`,
    );
  }
}

File: python/src/contriwork_config_core/resolvers.py
"""SecretResolver protocol and built-in implementations.

See CONTRACT.md §Port/`SecretResolver protocol`. Each resolver handles one or
more schemes of ``${scheme:value}``. Unknown schemes raise
:class:`SecretSchemeUnsupported`; recognized-but-unresolvable values raise
:class:`SecretRefUnresolved`. :class:`ChainResolver` composes several.
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Protocol, runtime_checkable

from .errors import SecretRefUnresolved, SecretSchemeUnsupported


@runtime_checkable
class SecretResolver(Protocol):
    """Resolve a ``${scheme:value}`` reference to a plain string."""

    async def resolve(self, scheme: str, value: str) -> str:
        """Return the resolved secret or raise.

        - :class:`SecretSchemeUnsupported` if this resolver does not handle
          ``scheme``. :class:`ChainResolver` relies on this signal to try
          the next resolver.
        - :class:`SecretRefUnresolved` if ``scheme`` is handled but ``value``
          cannot be produced (env var not set, file missing, etc.).
        """
        ...


class EnvResolver:
    """Resolve ``${env:NAME}`` from the process environment."""

    scheme = "env"

    async def resolve(self, scheme: str, value: str) -> str:
        if scheme != self.scheme:
            raise SecretSchemeUnsupported(f"EnvResolver does not handle scheme {scheme!r}")
        if not value:
            raise SecretRefUnresolved("${env:} requires a variable name")
        resolved = os.environ.get(value)
        if resolved is None:
            raise SecretRefUnresolved(f"env var not set: {value}")
        return resolved


class FileResolver:
    """Resolve ``${file:path}`` by reading the file's contents.

    Trailing whitespace (incl. newline) is stripped from the result — this
    matches the common "secret in a single-line file" pattern used by
    Docker secrets and ``/run/secrets/``. If ``base_dir`` is provided,
    relative paths resolve against it; absolute paths are honored as-is.
    """

    scheme = "file"

    def __init__(self, base_dir: str | Path | None = None) -> None:
        self._base_dir = Path(base_dir) if base_dir is not None else None

    async def resolve(self, scheme: str, value: str) -> str:
        if scheme != self.scheme:
            raise SecretSchemeUnsupported(f"FileResolver does not handle scheme {scheme!r}")
        if not value:
            raise SecretRefUnresolved("${file:} requires a path")
        path = Path(value)
        if not path.is_absolute() and self._base_dir is not None:
            path = self._base_dir / path
        try:
            text = path.read_text(encoding="utf-8")
        except FileNotFoundError:
            raise SecretRefUnresolved(f"secret file not found: {path}") from None
        except OSError as e:
            raise SecretRefUnresolved(f"cannot read secret file {path}: {e}") from e
        return text.rstrip()


class ChainResolver:
    """Try each resolver in order; first one that handles the scheme wins.

    - If a resolver raises :class:`SecretSchemeUnsupported`, move on.
    - Any other exception propagates immediately (including
      :class:`SecretRefUnresolved`).
    - If no resolver claims the scheme, raise :class:`SecretSchemeUnsupported`
      naming the scheme.
    """

    def __init__(self, *resolvers: SecretResolver) -> None:
        self._resolvers = resolvers

    async def resolve(self, scheme: str, value: str) -> str:
        for r in self._resolvers:
            try:
                return await r.resolve(scheme, value)
            except SecretSchemeUnsupported:
                continue
        raise SecretSchemeUnsupported(f"no resolver registered for scheme {scheme!r}")
